crates/core/tedge/src/cli/certificate/c8y/download.rs

```diff
@@ -0,0 +1,178 @@
+use crate::cli::certificate::c8y::create_device_csr;
+use crate::cli::certificate::c8y::read_csr_from_file;
+use crate::cli::certificate::c8y::store_device_cert;
+use crate::command::Command;
+use crate::error;
+use crate::get_webpki_error_from_reqwest;
+use crate::log::MaybeFancy;
+use crate::warning;
+use anyhow::Context;
+use anyhow::Error;
+use c8y_api::json_c8y_deserializer::C8yAPIError;
+use camino::Utf8PathBuf;
+use certificate::CloudRootCerts;
+use certificate::CsrTemplate;
+use hyper::StatusCode;
+use reqwest::header::CONTENT_TYPE;
+use reqwest::Response;
+use std::io::Write;
+use std::time::Duration;
+use tedge_config::models::HostPort;
+use tedge_config::models::HTTPS_PORT;
+use url::Url;
+
+/// Command to request and download a device certificate from Cumulocity
+pub struct DownloadCertCmd {
+    /// The device identifier to be used as the common name for the certificate
+    pub device_id: String,
+
+    /// The security token assigned to this device when registered to Cumulocity
+    pub security_token: String,
+
+    /// Cumulocity instance where the device has been registered
+    pub c8y_url: HostPort<HTTPS_PORT>,
+
+    /// Root certificates used to authenticate the Cumulocity instance
+    pub root_certs: CloudRootCerts,
+
+    /// The path where the device certificate will be stored
+    pub cert_path: Utf8PathBuf,
+
+    /// The path where the device private key will be stored
+    pub key_path: Utf8PathBuf,
+
+    /// The path where the device CSR file will be stored
+    pub csr_path: Utf8PathBuf,
+
+    /// Tell if the CSR has to be generated or is ready to be used
+    pub generate_csr: bool,
+
+    /// Delay between two attempts, polling till the device is registered
+    pub retry_every: Duration,
+
+    /// Maximum time waiting for the device to be registered
+    pub max_timeout: Duration,
+
+    /// CSR template
+    pub csr_template: CsrTemplate,
+}
+
+#[async_trait::async_trait]
+impl Command for DownloadCertCmd {
+    fn description(&self) -> String {
+        format!(
+            "Download a certificate from {} for the device {}",
+            self.c8y_url, self.device_id
+        )
+    }
+
+    async fn execute(&self) -> Result<(), MaybeFancy<Error>> {
+        Ok(self.download_device_certificate().await?)
+    }
+}
+
+impl DownloadCertCmd {
+    async fn download_device_certificate(&self) -> Result<(), Error> {
+        let (common_name, security_token) = self.get_registration_data()?;
```

Comment:
This should probably use `spawn_blocking`

Reply:
This has been addressed by https://github.com/thin-edge/thin-edge.io/pull/3525/commits/ec2bf687f3ded1a07585508a7bd7b838cc21a5ab